File: mnp/mnpgo/server/internal/stream_multiplexer.go
/*
 * This file is part of the Mantik Project.
 * Copyright (c) 2020-2021 Mantik UG (Haftungsbeschränkt)
 * Authors: See AUTHORS file
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License version 3.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.
 *
 * Additionally, the following linking exception is granted:
 *
 * If you modify this Program, or any covered work, by linking or
 * combining it with other code, such other code is not for that reason
 * alone subject to any of the requirements of the GNU Affero GPL
 * version 3.
 *
 * You can be released from the requirements of the license by purchasing
 * a commercial license.
 */
package internal

import (
	"bufio"
	"github.com/mantik-ai/core/mnp/mnpgo/protos/mantik/mnp"
	"io"
	"sync"
)

// A Buffered Stream Multiplexer/Demultiplexer
// Used for handling different input/output channels into a group of readers and writers
// Readers and writes are generated using pipes

type portInfo struct {
	err   error
	calls int
	bytes int64
}

type StreamMultiplexer struct {
	bufSize int

	inputStates  []portInfo
	outputStates []portInfo
	mutex        sync.Mutex

	inputs       []io.WriteCloser
	InputReaders []io.Reader
	outputs      []io.Reader
	OutputWrites []io.WriteCloser
}

func NewStreamMultiplexer(inputChannels int, outputChannels int) *StreamMultiplexer {
	bufSize := 65536 // Configurable?

	r := StreamMultiplexer{
		bufSize:      bufSize,
		inputStates:  make([]portInfo, inputChannels, inputChannels),
		outputStates: make([]portInfo, outputChannels, outputChannels),
		inputs:       make([]io.WriteCloser, inputChannels, inputChannels),
		InputReaders: make([]io.Reader, inputChannels, inputChannels),
		outputs:      make([]io.Reader, outputChannels, outputChannels),
		OutputWrites: make([]io.WriteCloser, outputChannels, outputChannels),
	}

	for i := 0; i < inputChannels; i++ {
		inputReader, inputWriter := io.Pipe()
		bufferedInputReader := bufio.NewReaderSize(inputReader, bufSize)
		r.inputs[i] = inputWriter
		r.InputReaders[i] = bufferedInputReader
	}

	for i := 0; i < outputChannels; i++ {
		outputReader, outputWriter := io.Pipe()
		bufferedOutputReader := bufio.NewReaderSize(outputReader, bufSize)
		r.outputs[i] = bufferedOutputReader
		r.OutputWrites[i] = outputWriter
	}
	return &r
}

func (s *StreamMultiplexer) InputReader(id int) io.Reader {
	return s.InputReaders[id]
}

func (s *StreamMultiplexer) OutputWriter(id int) io.Writer {
	return s.OutputWrites[id]
}

// Write something on the input side
func (s *StreamMultiplexer) Write(id int, bytes []byte) error {
	_, err := s.inputs[id].Write(bytes)
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.inputStates[id].calls += 1
	s.inputStates[id].bytes += (int64)(len(bytes))
	if err != nil {
		s.inputStates[id].err = err
	}
	return err
}

func (s *StreamMultiplexer) WriteEof(id int) error {
	err := s.inputs[id].Close()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	state := err
	if state == nil {
		state = io.EOF
	}
	s.inputStates[id].err = state
	return err
}

func (s *StreamMultiplexer) WriteFailure(id int, err error) {
	s.inputs[id].Close()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.inputStates[id].err = err
}

func (s *StreamMultiplexer) ReadFailure(id int, err error) {
	s.OutputWrites[id].Close()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.outputStates[id].err = err
}

// Read something on the output side
func (s *StreamMultiplexer) Read(id int) ([]byte, error) {
	buffer := make([]byte, s.bufSize, s.bufSize)

	n, err := io.ReadFull(s.outputs[id], buffer)

	s.mutex.Lock()
	defer s.mutex.Unlock()

	s.outputStates[id].calls += 1

	if n <= 0 && err != nil {
		if err == io.ErrUnexpectedEOF {
			err = io.EOF
		}
		if s.outputStates[id].err == nil {
			s.outputStates[id].err = err
		} else {
			err = s.outputStates[id].err
		}
		return []byte{}, err
	}
	s.outputStates[id].bytes += (int64)(n)
	return buffer[:n], nil
}

func (s *StreamMultiplexer) Finalize(err error) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	for i, inputState := range s.inputStates {
		if inputState.err == nil {
			s.inputStates[i].err = err
			s.inputs[i].Close()
		}
	}
	for i, outputState := range s.outputStates {
		if outputState.err == nil {
			s.outputStates[i].err = err
			s.OutputWrites[i].Close()
		}
	}
}

// Returns port status for input and ouptut
func (s *StreamMultiplexer) QueryTaskPortStatus() ([]*mnp.TaskPortStatus, []*mnp.TaskPortStatus) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	return convertStates(s.inputStates), convertStates(s.outputStates)
}

func convertStates(info []portInfo) []*mnp.TaskPortStatus {
	result := make([]*mnp.TaskPortStatus, len(info), len(info))
	for i, v := range info {
		state := convertState(v)
		result[i] = &state
	}
	return result
}

func convertState(info portInfo) mnp.TaskPortStatus {
	var result mnp.TaskPortStatus
	if info.err != nil && info.err != io.EOF {
		result.Error = info.err.Error()
	}
	result.MsgCount = (int32)(info.calls)
	result.Data = info.bytes
	result.Done = info.err == io.EOF
	return result
}

File: mnp/mnppython/mnp/streammultiplexer.py
#
# This file is part of the Mantik Project.
# Copyright (c) 2020-2021 Mantik UG (Haftungsbeschränkt)
# Authors: See AUTHORS file
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License version 3.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.
#
# Additionally, the following linking exception is granted:
#
# If you modify this Program, or any covered work, by linking or
# combining it with other code, such other code is not for that reason
# alone subject to any of the requirements of the GNU Affero GPL
# version 3.
#
# You can be released from the requirements of the license by purchasing
# a commercial license.
#

from typing import List
from typing import BinaryIO
from mnp._stubs.mantik.mnp.mnp_pb2 import TaskPortStatus
import io

import os

# Max Chunk size to aovid deadlock via Pipes
MAX_CHUNK_SIZE = 8192


class StreamyPipe:

    def __init__(self):
        read_fd, write_fd = os.pipe()
        self.read_fd = read_fd
        self.write_fd = write_fd
        self.writer = io.open(write_fd, "wb")
        self.reader = io.open(read_fd, "rb")
        self.calls = 0
        self.byte_count = 0
        self.is_done = False

    def write(self, b: bytes):
        """
        Write something on the incoming side
        """
        result = self.writer.write(b)
        self.calls += 1
        self.byte_count += len(b)
        return result

    def write_eof(self):
        """
        Close incoming side
        """
        result = self.writer.close()
        self.calls += 1
        self.is_done = True
        return result

    def read(self) -> bytes:
        """
        Read out side
        """
        result = self.reader.read(MAX_CHUNK_SIZE)
        self.calls += 1
        self.byte_count += len(result)
        if not result:
            self.is_done = True
        return result

    def task_port_status(self) -> TaskPortStatus:
        # No error support here
        return TaskPortStatus(
            msg_count=self.calls,
            data=self.byte_count,
            done=self.is_done
        )


class StreamMultiplexer:
    """
    Handles input and output streams of a Task
    """

    inputs: List[StreamyPipe]
    outputs: List[StreamyPipe]

    def __init__(self, input_num: int, output_num: int):
        self.inputs = [StreamyPipe() for i in range(input_num)]
        self.outputs = [StreamyPipe() for i in range(output_num)]

    def wrapped_inputs(self) -> List[BinaryIO]:
        return [i.reader for i in self.inputs]

    def wrapped_outputs(self) -> List[BinaryIO]:
        return [i.writer for i in self.outputs]

    def push(self, port: int, b: bytes):
        return self.inputs[port].write(b)

    def push_eof(self, port: int):
        return self.inputs[port].write_eof()

    def pull(self, port: int) -> bytes:
        return self.outputs[port].read()

    def output_reader(self, port: int) -> BinaryIO:
        return self.outputs[port].reader

    def close_output(self):
        for output in self.outputs:
            output.writer.close()

    def close(self):
        for output in self.outputs:
            output.writer.close()
        for input in self.inputs:
            input.writer.close()

    def input_port_status(self) -> List[TaskPortStatus]:
        """
        Reports input port status in an MNP Compatible way
        """
        return list(map(StreamyPipe.task_port_status, self.inputs))

    def output_port_status(self) -> List[TaskPortStatus]:
        """
        Report output port status in a MNP Compatible way.
        """
        return list(map(StreamyPipe.task_port_status, self.outputs))